### PR TITLE
Get release time

### PR DESCRIPTION
**Pull request by KPostOffice**
Get the date and time of the release. Useful data analysis with respect to time.

### CODE REVIEW

File: thoth/python/source.py [hunk 1]
```diff
@@ -405,3 +406,15 @@ def get_package_hashes(self, package_name: str, package_version: str, with_inclu
             result.append(doc)
 
         return result
+
+    def get_package_release_date(self, package_name: str, package_version: str,) -> datetime:
+        """Get time at which package was uploaded to package index."""
+        if not self.warehouse:
+            raise NotImplementedError("Cannot release date for non-warehouse repository.")
+        package_json = requests.get(f"{self.get_api_url()}/{package_name}/json").json()
```
Comment (pacospace) [inline thoth/python/source.py:414]: What about reusing logic already here: https://github.com/thoth-station/python/blob/dc489d8fbb56702f57e14c4a053e70f2ccae877e/thoth/python/source.py#L120

```python
    def _warehouse_get_api_package_info(self, package_name: str) -> dict:
        """Use API of the deployed Warehouse to gather package information."""
        url = self.get_api_url() + f"/{package_name}/json"
        _LOGGER.debug("Gathering package information from Warehouse API: %r", url)
        response = requests.get(url, verify=self.verify_ssl)
        if response.status_code == 404:
            raise NotFound(f"Package {package_name} not found on warehouse {self.url} ({self.name})")
        response.raise_for_status()
        return response.json()
```

File: thoth/python/source.py [hunk 2]
```diff
@@ -405,3 +406,13 @@ def get_package_hashes(self, package_name: str, package_version: str, with_inclu
             result.append(doc)
 
         return result
+
+    def get_package_release_date(self, package_name: str, package_version: str,) -> datetime:
+        """Get time at which package was uploaded to package index."""
+        package_json = self._warehouse_get_api_package_info(package_name)
+        release = package_json["releases"][package_version]
```
Comment (fridex) [inline thoth/python/source.py:413]: It might be good idea to raise `NotFound` if the given version was not found.

File: thoth/python/source.py [hunk 3]
```diff
@@ -405,3 +406,13 @@ def get_package_hashes(self, package_name: str, package_version: str, with_inclu
...
+        artifact = next(x for x in release if x["python_version"] == "source")
+        if artifact is None:
+            raise ValueError(f"No source distribution found on {self.warehouse_api_url}")
```
Comment (fridex) [inline thoth/python/source.py:416]: You can also add name and version of the package so the exception is self-descriptive and easily identified when encountered in the cluster.

File: thoth/python/source.py [hunk 4]
```diff
@@ -405,3 +406,13 @@ def get_package_hashes(self, package_name: str, package_version: str, with_inclu
...
+
+        return datetime.strptime(artifact["upload_time"], "%Y-%m-%dT%H:%M:%S")
```
Comment (fridex) [inline thoth/python/source.py:418]: We have methods for manipulating with dates and datetimes in `thoth.common.helpers`. It helps to keep track of timezone issues and date/datetime formatting across the whole Thoth application.

File: thoth/python/source.py (hunk 3 above)
Comment (fridex) [reply thoth/python/source.py:416]: Also, you can use thoth-python exception `NotFound`, so anyone using thoth-python can handle library exceptions (respecting exception hierarchy).

**Comment by goern** [comment]
recheck

**Pull request merged**